### ISSUE TITLE
scipy.ndimage.find_objects

### ISSUE THREAD

**Issue by kazemSafari** [pytorch/pytorch#102201]
### 🚀 The feature, motivation and pitch

This function a basic building block of any biomedical image analysis application.
It gives a list of tuple of slices of coordinates of labelled objects within a mask image of dtype Uint16 or Uint8, assuming the image
background is 0, and the labelled objects go from 1, 2, ..., max_label.

I was wondering it is possible to implement it in torch C++ using a simple TensorIterator.

Basically the simplest case would be it takes a 2D tensor of size (H, W) as input and
outputs a tensor of slices of size (N, 2) where N is the number of objects, and each row
is [slice(start,end,step), slice(start,end,step)]. 

### Alternatives

The implementation in C numpy can be found here:
https://github.com/scipy/scipy/blob/v1.10.1/scipy/ndimage/src/ni_measure.c

which uses Iterators defined here:
https://github.com/scipy/scipy/blob/v1.10.1/scipy/ndimage/src/ni_support.h

### Additional context

Can it also be extended to allow extract objects from a tensor of dimension (B, C, W, H) where B is the batch size, C the number of
channels and W is the width and H is the height.

**Comment by albanD** [comment]
cc @rgommers in case you have ideas!

**Comment by soulitzer** [comment]
Thanks for the proposal. I was wondering how this fits into a image segmentation workflow in general, specifically why not have whatever image segmentation model you are running to just generate the masks separately - isn't that how its usually done?

**Comment by kazemSafari** [comment]
![w1_cdp2bioactives_a01_s1](https://github.com/pytorch/pytorch/assets/32110820/d8757f25-6d59-4a68-95a6-27a9f7a41d81)
![w2_cdp2bioactives_a01_s1](https://github.com/pytorch/pytorch/assets/32110820/1cff8ef9-5e87-4d09-96fe-8df01fe76bcb)
![w3_cdp2bioactives_a01_s1](https://github.com/pytorch/pytorch/assets/32110820/a78123a0-b302-466c-8b2a-3ca894912202)
![w4_cdp2bioactives_a01_s1](https://github.com/pytorch/pytorch/assets/32110820/6c458508-9ecc-47d8-8c76-e0ee7eebd0d5)
![w5_cdp2bioactives_a01_s1-1](https://github.com/pytorch/pytorch/assets/32110820/340a507d-959d-487f-b65a-b6ff6bd54af1)

**Comment by kazemSafari** [comment]
![w1_cdp2bioactives_a01_s1](https://github.com/pytorch/pytorch/assets/32110820/cc7cf979-73c0-41f7-81b6-202f0f60243c)

**Comment by soulitzer** [comment]
Thanks for the context. If you do not need to backprop through find_objects with autograd, it seems that the workaround here is to just convert to numpy and back here.

**Comment by kazemSafari** [comment]
That is correct. 

But one still will have to do a triple for loop in python ( to extract from each image, each object/cell, from each channel) which is very slow because of GIL. I am using the multiprocessing module over the images.

Also, the find_objects function only gives the coordinates of the objects. It does not extract them in a python list/ or a C++ vector. So I can't readily convert them to a tensor.

**Comment by kazemSafari** [comment]
@soulitzer @albanD @jbschlosser
> Thanks for the context. If you do not need to backprop through find_objects with autograd, it seems that the workaround here is to just convert to numpy and back here.
I just found [about ](https://pytorch.org/vision/main/auto_examples/plot_repurposing_annotations.html#sphx-glr-download-auto-examples-plot-repurposing-annotations-py) which is exactly what i need. 
I am sorry because I did not know about it before I started this thread.

But the only problem is it does not support tiff files format yet.
Can you add support for reading tiff files to torchvision.io.read_image?
Thanks.

**Comment by soulitzer** [comment]
Oh interesting, I saw that too, but they seemed different to me because masks_to_boxes just computes the min/max values of a single provided mask, whereas find_objects takes a single image with more than one mask and extracts out the connected components.

Btw for those requests you may want to post to https://github.com/pytorch/vision instead?

**Comment by rgommers** [comment]
> It is slower in torch maybe due to that fact the pytorch does not support uint16 natively.

This is the feature request, with a number of image processing use cases and a thumbs up about supporting it now that the PyTorch 2.0 infrastructure makes it easier to do so: gh-58734.

>> Btw for those requests you may want to post to https://github.com/pytorch/vision instead?

> Yeah sure. All three of them?!

I believe so, yes - all `scipy.ndimage` equivalents probably belong in `torchvision`. 

Given the performance results you posted above, even the GPU version of `torchvision.ops.masks_to_boxes` is much slower than SciPy though, so there's probably not much point in using `masks_to_boxes` unless the performance can be improved.

**Comment by vadimkantorov** [comment]
Somewhat related:
- https://github.com/pytorch/vision/issues/4415
